File: des/bin/checksum-compare.py
"""
 Compare our checksum list on files to the master list
"""
import sys
def load_files(fname, is_master=False):
    if is_master:
        namestart=2
    else:
        namestart=27

    flist={}
    with open(fname) as fobj:
        for line in fobj:
            md5sum,name=line.split()
            name=name[namestart:]

            flist[name] = md5sum
    return flist        

def compare_md5sums(master_flist, test_flist):
    for name,md5sum in test_flist.iteritems():
        try:
            master_md5sum=master_flist[name]
            if md5sum != master_md5sum:
                print name,'has md5sum',md5sum,'instead of',master_md5sum
        except:
            print '%s not found in master list' % name
def main():
    if len(sys.argv) != 3:
        print "usage: checksum-compare.py master_list test_list"
        sys.exit(45)

    master_file=sys.argv[1]
    test_file=sys.argv[2]

    master_flist = load_files(master_file, is_master=True)
    test_flist = load_files(test_file, is_master=False)

    compare_md5sums(master_flist, test_flist)

main()

File: des/src/checksum-compare.c
/*
 * Compare our checksum list on files to the master list
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "uthash.h"

#define FLEN 255
#define MD5LEN 33
#define MASTER_NAMESTART 2
#define TEST_NAMESTART 27

struct filehash {
    char name[FLEN];
    char md5sum[MD5LEN];
    UT_hash_handle hh; /* makes this structure hashable */
};

FILE* file_open(char* filename) {
    FILE* fobj=fopen(filename,"r");
    if (!fobj) {
        fprintf(stderr,"Could not open file: %s\n", filename);
        exit(45);
    }
    return fobj;
}

struct filehash* load_master(char* filename) {
    char name[FLEN], md5sum[MD5LEN];
    struct filehash* files=NULL;
    int namestart=2;

    FILE* fobj=file_open(filename);

    while ((fscanf(fobj, "%s %s", md5sum, name) == 2)) {
        struct filehash* thisfile=calloc(1,sizeof(struct filehash));

        strcpy(thisfile->name, name+MASTER_NAMESTART);
        strcpy(thisfile->md5sum, md5sum);
        HASH_ADD_STR(files, name, thisfile);
    }
    fclose(fobj);

    return files;
}

struct filehash* find_file(struct filehash* files, char* name) {
    // a single file reference, don't allocate
    struct filehash* afile=NULL;
    HASH_FIND_STR(files, name, afile);
    return afile;
}

void compare_md5sums(struct filehash* master_files, char* filename) {
    struct filehash* master_file=NULL;
    char name[FLEN], md5sum[MD5LEN];

    FILE* fobj=file_open(filename);

    while ((fscanf(fobj, "%s %s", md5sum, name) == 2)) {
        master_file=find_file(master_files, name+TEST_NAMESTART);
        if (master_file==NULL) {
            fprintf(stderr,"%s not found in master list\n", name);
        } else {
            if (strcmp(md5sum, master_file->md5sum) != 0) {
                fprintf(stderr,"%s has md5sum %s instead of %s\n", 
                               name, md5sum, master_file->md5sum);
            }
        }
    }
}
int main(int argc, char** argv) {
    if (argc < 3) {
        puts("usage: checksum-compare master_list test_list\n");
        exit(45);
    }

    // this hash table will represent our files/md5sums, keyed by name
    struct filehash* master_files=NULL;

    master_files = load_master(argv[1]);
    compare_md5sums(master_files, argv[2]);

    return 0;
}
